File: OnlyCueTests/Support/GoldenDouble.swift
import Foundation

/// A `Double` carried through a golden vector as a **round-trip-exact decimal
/// string** rather than a JSON number.
///
/// The M1 contract (`docs/superpowers/specs/2026-09-14-windows-m1-contract-design.md`)
/// requires this because Swift's and .NET's shortest-representation formatters do
/// not spell the same value the same way — Swift writes `1.0` and `1e-05` where
/// .NET's `"R"` writes `1` and `1E-05`. Comparing the *text* would therefore
/// report drift that isn't there. The string is only transport: both cores parse
/// it back to an IEEE-754 `double` and compare bit patterns.
///
/// Bit-pattern equality (not `==`) is deliberate. It makes `-0.0` distinct from
/// `0.0` and `NaN` equal to itself, so a vector can pin those cases instead of
/// silently passing them.
///
/// NaN travels as `nan:0x<16 hex digits>` rather than as a mnemonic, because a
/// mnemonic cannot name it (#833). Measured on both runtimes: Swift's
/// `Double.nan` — and every NaN that comes out of widening a binary32 NaN — is
/// `0x7FF8000000000000`, while .NET's `double.NaN` is `0xFFF8000000000000`, and
/// .NET's `TryParse` maps *both* `"nan"` and `"-nan"` onto the sign-bit-set
/// pattern. So no text .NET understands can name the value Swift writes, and the
/// bitwise comparison this contract runs on would report drift that isn't there.
/// The `nan:` marker is not decoration: `Double("0x7FF8000000000000")` is
/// **9.221120237041091e+18** on Swift — a plausible finite number, silently —
/// where .NET refuses it, so a bare-hex spelling would have been the one carrier
/// worse than the mnemonic. `inf` / `-inf` have a single bit pattern each and are
/// left as they are.
struct GoldenDouble: Codable, Equatable, ExpressibleByFloatLiteral, ExpressibleByIntegerLiteral {

    /// What a NaN's bit pattern is prefixed with in the vector text.
    private static let nanMarker = "nan:0x"

    /// The characters Swift's `description` can emit for a finite `Double`.
    /// Anything else — a hex float, a mnemonic, stray whitespace — is refused
    /// rather than handed to `Double(_:)`, whose acceptance differs from .NET's
    /// in both directions.
    private static let decimalCharacters = Set("0123456789+-.eE")

    let value: Double

    init(_ value: Double) { self.value = value }

    init(floatLiteral value: Double) { self.init(value) }

    init(integerLiteral value: Int) { self.init(Double(value)) }

    init(from decoder: Decoder) throws {
        let text = try decoder.singleValueContainer().decode(String.self)
        guard let parsed = Self.parse(text) else {
            throw DecodingError.dataCorruptedError(
                in: try decoder.singleValueContainer(),
                debugDescription: "\(text) is not a parseable Double"
            )
        }
        self.init(parsed)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        // Swift's `description` is the shortest string that round-trips — except
        // for NaN, where it is a mnemonic (`nan`, or `nan(0x1)` for a payload,
        // which `Double(_:)` cannot even read back).
        guard !value.isNaN else {
            try container.encode(Self.nanMarker + String(format: "%016llX", value.bitPattern))
            return
        }
        try container.encode("\(value)")
    }

    private static func parse(_ text: String) -> Double? {
        if let hex = text.hasPrefix(nanMarker) ? text.dropFirst(nanMarker.count) : nil {
            guard hex.count == 16, let bits = UInt64(hex, radix: 16) else { return nil }
            return Double(bitPattern: bits)
        }
        if text == "inf" { return .infinity }
        if text == "-inf" { return -.infinity }
        guard !text.isEmpty, text.allSatisfy(decimalCharacters.contains) else { return nil }
        return Double(text)
    }

    static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.value.bitPattern == rhs.value.bitPattern
    }
}

extension GoldenDouble {
    /// Lifts an optional `Double` without spelling the `map` at every call site.
    init?(_ value: Double?) {
        guard let value else { return nil }
        self.init(value)
    }
}

File: windows/OnlyCue.Core.Tests/GoldenDouble.cs
using System.Globalization;

namespace OnlyCue.Core.Tests;

/// <summary>
/// Reads the M1 contract's round-trip-exact decimal strings back into IEEE-754
/// doubles, and compares them the way the contract specifies: by bit pattern.
/// </summary>
/// <remarks>
/// The vectors carry doubles as strings rather than JSON numbers because Swift's
/// and .NET's shortest-representation formatters spell the same value
/// differently (<c>1.0</c> vs <c>1</c>, <c>1e-05</c> vs <c>1E-05</c>), which would
/// read as drift that isn't there. The string is transport only — see
/// <c>OnlyCueTests/Support/GoldenDouble.swift</c> for the writing half.
///
/// Bit-pattern comparison (not <c>==</c>) is deliberate: it keeps <c>-0.0</c>
/// distinct from <c>0.0</c> and makes <c>NaN</c> equal itself, so a vector can
/// pin those instead of silently passing them.
///
/// NaN arrives as <c>nan:0x&lt;16 hex digits&gt;</c> rather than as a mnemonic,
/// because no text this runtime understands can name the NaN Swift writes
/// (#833). Measured: Swift's <c>Double.nan</c>, and every NaN obtained by
/// widening a binary32 NaN, is <c>0x7FF8000000000000</c>; <c>double.NaN</c> here
/// is <c>0xFFF8000000000000</c>, and <c>TryParse</c> maps <em>both</em>
/// <c>"nan"</c> and <c>"-nan"</c> onto that sign-bit-set pattern. So the
/// mnemonics are now refused rather than reinterpreted — accepting one would mean
/// choosing a sign on the writer's behalf.
/// </remarks>
internal static class GoldenDouble
{
    private const string NanMarker = "nan:0x";

    /// <summary>The characters Swift's <c>description</c> emits for a finite
    /// double. Restricting the fallback to them refuses, in one place, every
    /// spelling exactly one platform accepts: <c>"NaN"</c>/<c>"Infinity"</c>
    /// (.NET only) and hex floats like <c>0x1p3</c> (Swift only, where it is
    /// silently 8.0).</summary>
    private const string DecimalCharacters = "0123456789+-.eE";

    public static double Parse(string text)
    {
        if (text.StartsWith(NanMarker, StringComparison.Ordinal))
        {
            var hex = text.AsSpan(NanMarker.Length);
            return hex.Length == 16 && long.TryParse(hex, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var bits)
                ? BitConverter.Int64BitsToDouble(bits)
                : throw new InvalidDataException($"'{text}' is not a 64-bit NaN pattern");
        }

        // The infinities keep their mnemonics — one bit pattern each, so nothing
        // is lost — and .NET's invariant culture does not spell them this way.
        switch (text)
        {
            case "inf": return double.PositiveInfinity;
            case "-inf": return double.NegativeInfinity;
        }

        if (text.Length == 0 || text.AsSpan().IndexOfAnyExcept(DecimalCharacters) >= 0)
        {
            throw new InvalidDataException($"'{text}' is not a spelling the contract writes");
        }

        return double.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out var value)
            ? value
            : throw new InvalidDataException($"'{text}' is not a parseable double");
    }

    /// <summary>The optional form. A <c>null</c> in the vector is a real result —
    /// "no grid line here", "this cue was left unassigned" — not a missing case.</summary>
    public static double? ParseOrNull(string? text) => text is null ? null : Parse(text);

    public static bool BitwiseEquals(double expected, double actual) =>
        BitConverter.DoubleToInt64Bits(expected) == BitConverter.DoubleToInt64Bits(actual);

    public static bool BitwiseEquals(double? expected, double? actual) =>
        (expected, actual) switch
        {
            (null, null) => true,
            ({ } lhs, { } rhs) => BitwiseEquals(lhs, rhs),
            _ => false
        };

    /// <summary>A description that distinguishes values the shortest form hides —
    /// used only in assertion messages.</summary>
    public static string Describe(double? value) =>
        value is { } number
            ? $"{number.ToString("R", CultureInfo.InvariantCulture)} (0x{BitConverter.DoubleToInt64Bits(number):X16})"
            : "null";
}
